### PR TITLE
fix: keep streamed answers out of thinking and improve local bash probes

### LINKED ISSUE #4997: [bug]  前端输出的正文在思考框内
### Before you start

- [x] I searched [existing issues](https://github.com/bytedance/deer-flow/issues?q=is%3Aissue) and this is not a duplicate.
- [x] I can reproduce this on the latest `main`.

### Problem summary

前端输出的正文在思考框内

### Affected area(s)

Frontend (UI / Next.js)

### What happened?

发起问答后，最终回答的正文会先输出在思考框内，输出完后，思考框内的正文会突然变到外面，这种情况只是看起来有点奇怪，但并没有实际的影响，不过理论上，正文不应该输出到思考框内。

但有时候思考框内的正文不会变到外面，问答结束后仍然停留在思考框内，这种情况只是偶尔出现，执行长任务的时候概率高一些。如下图所示。

local和docker环境都会出现这种情况。

我的模型配置是：
   - name: deepseek-v4-flash
     display_name: deepseek-v4-flash
     use: deerflow.models.patched_deepseek:PatchedChatDeepSeek
     model: deepseek-v4-flash
     api_base: https://dashscope.aliyuncs.com/compatible-mode/v1
     api_key: 
     timeout: 600.0
     max_retries: 10
    #  max_tokens: 20192
     supports_thinking: true
     supports_vision: false  # DeepSeek V3 does not support vision
     when_thinking_enabled:
       extra_body:
         thinking:
           type: enabled

Image

### Expected behavior

.

### Steps to reproduce

.

### Relevant logs

```shell
.
```

### How are you running DeerFlow?

Docker (make docker-start)

### Operating system

Linux

### Platform details

_No response_

### Python version

_No response_

### Node.js version

_No response_

### pnpm version

_No response_

### uv version

_No response_

### Git state

_No response_

### Support bundle summary

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by nothingyuancando**
Fixes #4997 and #4999

## Why

Streaming answer text from reasoning-capable providers could be grouped into the processing disclosure until a turn settled, so it appeared in the Thinking panel and could remain there after a long run. Separately, the bash tool advertised a Linux-only environment despite being available against an opted-in local host, which encouraged blocked Linux path probes and prevented useful local-system answers.

## What changed

- Streaming AI messages that already include reasoning plus answer text now render as an assistant bubble immediately; plain pre-tool narration remains in the processing group until the stream resolves.
- The model-visible bash contract now explains cross-platform local environment discovery: begin with `uname -s`, use `sw_vers` on Darwin, and recover from blocked absolute paths with command-only probes. The local path guard remains unchanged.

## Surface area

- [x] **Frontend UI** — streamed message grouping under `frontend/`
- [ ] **Backend API** — endpoint / SSE event / request-response shape under `backend/app`
- [ ] **Agents / LangGraph** — agent node, graph wiring, `langgraph.json`, or prompt change
- [x] **Sandbox** — sandboxed execution tool guidance
- [ ] **Skills** — change under `skills/`
- [ ] **Dependencies** — new/upgraded entry in `backend/pyproject.toml` or `frontend/package.json`
- [x] **Default behavior change** — streamed answer placement and bash tool-selection guidance
- [ ] **Docs / tests / CI only** — no runtime behavior change

## Screenshots / Recording

Not applicable. The frontend change is a deterministic message-grouping correction covered by unit tests; the other change is model-visible tool guidance.

## Bug fix verification

- `frontend/tests/unit/core/messages/utils.test.ts` reproduces #4997 with a live-turn AI message containing both `reasoning_content` and answer text. It grouped the answer as `assistant:processing` before this change; it now yields `assistant`.
- `backend/tests/test_local_sandbox_command_timeout.py` pins #4999 guidance for `uname -s`, `sw_vers`, Linux-neutral wording, and recovery without repeating a path rejected by the existing guard.

## Validation

- `cd frontend && python ../scripts/pnpm.py test -- tests/unit/core/messages/utils.test.ts tests/unit/components/workspace/messages/message-group.test.ts`
  - 1,018 passed; 0 skipped; 0 warnings
- `cd frontend && python ../scripts/pnpm.py format`
- `cd frontend && python ../scripts/pnpm.py typecheck`
- `cd frontend && python ../scripts/pnpm.py lint`
- `cd backend && uv run pytest tests/test_local_sandbox_command_timeout.py tests/test_sandbox_tools_security.py tests/test_local_bash_tool_loading.py -q`
  - 167 passed, 1 skipped
- `cd backend && make lint`

Not covered: real model tool-selection behavior and live external model/provider streaming; both need configured credentials and are outside deterministic unit coverage.

## AI assistance

**Tool(s) used:** OpenAI Codex

**How you used it:** Codex inspected the existing issue reports and open PRs to avoid duplicates, implemented the two targeted fixes, added regression tests, and ran the listed local validation. I reviewed the final diff and test results.

- [x] I've read and understand every line of this change and take responsibility for it — it's not unreviewed AI output.

### CODE REVIEW

**Comment by PiedPiper911** [comment]
Nice fix — I traced the grouping logic and the new `!hasReasoning(message)` guard is the right call. The three-way split is coherent:

- **reasoning + content** → assistant bubble (completing an answer, stays out of the thinking panel — fixes #4997)
- **reasoning only** (no content yet) → still lands in the processing group via `hasReasoning`
- **content only, no tool calls** → still `isUnresolvedAssistantText` (preserves the #4304 behavior so pre-tool narration doesn't jump)

Since the split is re-derived every render, this also handles the "long-run leftover" case where the answer text used to stay stuck in the thinking panel — good.

One suggestion: the new test only covers the **reasoning + content → assistant** path. It would be worth adding the mirror assertion — **reasoning-only (or reasoning + tool call) still stays in `assistant:processing`** — so future changes don't accidentally regress #3868 (reasoning painted twice) or #4304 (pre-tool text bouncing) while touching this condition. Cheap insurance for exactly the edge this PR is about.

**Comment by nothingyuancando** [comment]
@PiedPiper911 Thanks for the suggestion. I added the mirror regression coverage in 8deed94a: a streaming reasoning-only message is asserted to remain in `assistant:processing`, including the message membership assertion. The nearby existing test also continues to cover reasoning attached to a tool-calling step staying in the processing group. Verified with the focused message-utils suite (61 tests) and the full frontend `pnpm check`.

**Review by willem-bd — CHANGES_REQUESTED** [review]
I found two issues that should be addressed before merge:

1. **[P1] Update the streaming-order E2E contract.** The required E2E job on the previous head failed in `frontend/tests/e2e/streaming-reasoning-order.spec.ts`: after this change, the streamed answer is rendered in the assistant bubble, whose disclosure is labeled **Reasoning**, while the test still waits for **Thinking**. The refreshed head leaves that spec unchanged, so please update the E2E assertion/flow to match the intended UI and keep the required check green. Failure: https://github.com/bytedance/deer-flow/actions/runs/32726797868/job/97449062044

2. **[P2] Do not treat empty reasoning metadata as proof that answer text is final.** The new guard calls `hasReasoning(message)`, which returns true for any string-valued `additional_kwargs.reasoning_content`, including `""`. The MiMo adapter deliberately preserves empty strings, so a content-only pre-tool streaming message carrying `{ reasoning_content: "" }` bypasses `isUnresolvedAssistantText` and moves into the assistant bubble too early, regressing the pre-tool narration behavior this code preserves. Please require actual non-empty reasoning for this transition and add a regression test for content plus empty reasoning metadata.

Focused `messages/utils.test.ts` passes 61/61 on `db3bfde3`; the missing empty-metadata case is not covered.

**Comment by WillemJiang** [comment]
@nothingyuancando  please fix the e2e test error.

File: frontend/playwright.config.ts [hunk 1]
```diff
@@ -2,6 +2,7 @@ import { defineConfig, devices } from "@playwright/test";
 
 const baseURL = process.env.PLAYWRIGHT_BASE_URL ?? "http://localhost:3000";
 const skipWebServer = process.env.PLAYWRIGHT_SKIP_WEB_SERVER === "1";
+const useSystemChrome = process.env.PLAYWRIGHT_USE_SYSTEM_CHROME === "1";
```
Comment (willem-bd) [inline frontend/playwright.config.ts]: **[Suggestion] This change is unrelated to #4997/#4999 and undocumented.** The `PLAYWRIGHT_USE_SYSTEM_CHROME` opt-in isn't mentioned in the PR description or covered by any surface-area checkbox, yet this PR is labeled `risk:high` / `needs-validation`. It looks like local-tooling convenience - fine, but please either describe it in the PR body (why it's needed and when it's used) or split it into its own PR so this one stays a targeted two-fix change.

File: frontend/src/core/messages/utils.ts [hunk 2]
```diff
@@ -149,7 +149,11 @@ export function getMessageGroups(
         currentTurnStartIndex >= 0 &&
         messageIndex > currentTurnStartIndex &&
         hasContent(message) &&
-        !hasToolCalls(message);
+        !hasToolCalls(message) &&
+        // A provider that has already supplied reasoning with answer text is
+        // completing an answer, not merely streaming a pre-tool narration.
+        // Keep it out of the processing disclosure while the turn is active.
+        !hasReasoning(message);
       const becomesAssistantBubble =
```
Comment (willem-bd) [inline frontend/src/core/messages/utils.ts:161]: **[Question] Re-grouping edge for reasoning-bearing messages that later gain tool calls.** The #4304 behavior this condition preserves exists because providers can append tool-call chunks to a message that already streamed content - the text then jumps from the assistant bubble back into the processing group on the next render. With `!hasReasoning(message)` added here, a message carrying reasoning + answer text that later receives tool-call chunks (in the same merged AI message) now takes exactly that jump path: `becomesAssistantBubble` requires `!hasToolCalls`, so it re-enters the processing group mid-turn. Is the reasoning-then-content-then-tool-calls sequence ruled out for the providers in scope (the PR body's framing suggests reasoning+content is always terminal)? If it can happen, the existing unit tests would still pass but the visible jump regression would return for those providers - worth a line in the comment or a test pinning the assumption either way.

**Review by willem-bd — COMMENTED** [review]
Thanks - both items from my earlier review are addressed at this head: the E2E spec now expects the assistant bubble's "Reasoning" disclosure (label confirmed against `runDuration.reasoning` in `en-US.ts`), and `hasReasoning` now requires non-empty `reasoning_content` with regression tests for the empty-string and reasoning-only cases. Two smaller notes below: an undocumented, unrelated Playwright change riding along in this PR, and one question about the re-grouping edge for reasoning-bearing messages that later gain tool calls.

File: frontend/playwright.config.ts (hunk 1 above)
Comment (nothingyuancando) [reply frontend/playwright.config.ts]: Removed the unrelated `PLAYWRIGHT_USE_SYSTEM_CHROME` opt-in in a5193fc1. This PR now keeps its surface limited to the streaming-reasoning behavior and its E2E coverage.

File: frontend/src/core/messages/utils.ts (hunk 2 above)
Comment (nothingyuancando) [reply frontend/src/core/messages/utils.ts:161]: Pinned the re-grouping contract in a5193fc1. A reasoning-bearing content message initially renders as an assistant bubble; when its merged snapshot gains a tool call, it is deliberately reclassified as `assistant:processing` so the tool activity remains visible with the text that introduced it. The adjacent comment documents that choice, and the new regression test asserts both snapshots.

File: frontend/src/core/messages/utils.ts [hunk 3]
```diff
@@ -145,11 +145,19 @@ export function getMessageGroups(
...
       const isUnresolvedAssistantText =
         currentTurnStartIndex >= 0 &&
         messageIndex > currentTurnStartIndex &&
         hasContent(message) &&
-        !hasToolCalls(message);
+        !hasToolCalls(message) &&
+        // A provider that has already supplied reasoning with answer text is
+        // completing an answer, not merely streaming a pre-tool narration.
+        // Keep it out of the processing disclosure while the turn is active.
+        !hasReasoning(message);
```
Comment (willem-bd) [inline frontend/src/core/messages/utils.ts:160]: **[P2] Keep Anthropic thinking-only blocks in the processing group.** `hasContent()` returns true for any non-empty content array, including an Anthropic message whose only block is `{ type: "thinking" }`. Because that message also satisfies `hasReasoning()`, this new exception makes `isUnresolvedAssistantText` false and `becomesAssistantBubble` true while the turn is still loading. A diagnostic test with a human message followed by `content: [{ type: "thinking", thinking: "Still checking." }]` receives `["human", "assistant"]` instead of `["human", "assistant:processing"]`, so pure in-progress reasoning is exposed as an answer bubble for this supported provider shape. Please require actual non-reasoning answer content before applying the exception and add the array-content regression case alongside the new reasoning-only test.

**Review by willem-bd — CHANGES_REQUESTED** [review]
One provider-specific streaming regression remains at the current head. I reproduced it with a focused unit test; details are in the inline comment.

**Comment by WillemJiang** [comment]
@nothingyuancando  please take a look at the review comments.

**Comment by WillemJiang** [comment]
@nothingyuancando  please take a look at the review comments.

**Review by WillemJiang — APPROVED** [review]
